### PR TITLE
Cache SamplingRule.matches results

### PR DESCRIPTION
**Pull request by brettlangdon**
## Commit Message
{{title}}

Provided the same input Span service and operation names results from `SamplingRule.matches` will be consistent. To improve performance caching has been added to this method.

Sampling decisions generally only occur on local root spans, and _most_ applications only have a single local root span (1 service name, and 1 operation name) we can get a lot of benefits from caching the results.

## Performance results
A performance benchmark for `SamplingRule.matches` has been added to test the rule matching and performance under different cache hit/miss scenarios.

```
+------------------------------+-------------------------+------------------------------------------+
| Benchmark                    | ddtrace@master          | ddtrace@brettlangdon/cached.rule.matches |
+==============================+=========================+==========================================+
| samplingrules-high_match     | 107 us                  | 59.5 us: 1.80x faster                    |
+------------------------------+-------------------------+------------------------------------------+
| samplingrules-average_match  | 103 us                  | 59.8 us: 1.72x faster                    |
+------------------------------+-------------------------+------------------------------------------+
| samplingrules-low_match      | 97.8 us                 | 59.0 us: 1.66x faster                    |
+------------------------------+-------------------------+------------------------------------------+
| samplingrules-very_low_match | 951 us                  | 2.37 ms: 2.49x slower                    |
+------------------------------+-------------------------+------------------------------------------+
| Geometric mean               | (ref)                   | 1.20x faster                             |
+------------------------------+-------------------------+------------------------------------------+
```

The `very_low_rate` scenario is not surprising to be slower since the cache hit rate is very very low, and we run a lot of unique values through the cache so we hit the MFU cache limit (default is 256) and have to evict values from the cache. This is also a very highly unlikely user scenario. `high_match`, and `average_match` are closer to realistic scenarios with a decent cache hit rate of somewhere between 25% and 100%.

### CODE REVIEW

File: ddtrace/sampler.py [hunk 1]
```diff
@@ -396,6 +398,20 @@ def _pattern_matches(self, prop, pattern):
         # Exact match on the values
         return prop == pattern
 
+    @cachedmethod()
+    def _cached_matches(self, key):
```
Comment (P403n1x87) [inline ddtrace/sampler.py:402]:
```suggestion
    def _matches(self, key):
```

File: ddtrace/sampler.py [hunk 2]
```diff
@@ -406,13 +422,9 @@ def matches(self, span):
...
-        return all(
-            self._pattern_matches(prop, pattern)
-            for prop, pattern in [
-                (span.service, self.service),
-                (span.name, self.name),
-            ]
-        )
+        # Our MFU cache expects a single key, convert the
+        # provided Span into a hashable tuple for the cache
+        return self._cached_matches((span.service, span.name))
```
Comment (P403n1x87) [inline ddtrace/sampler.py:427]:
```suggestion
        return self._matches((span.service, span.name))
```

File: ddtrace/sampler.py [hunk 3]
```diff
@@ -24,6 +25,7 @@
 from .internal.compat import pattern_type
 from .internal.logger import get_logger
 from .internal.rate_limiter import RateLimiter
+from .internal.utils.cache import cachedmethod
```
Comment (P403n1x87) [inline ddtrace/sampler.py:28]: aside: the `cache` util module could be cythonised for even more performance?

**Review by P403n1x87 — COMMENTED** [review]
just a couple of suggestions, otherwise LGTM!

File: ddtrace/sampler.py (hunk 3 above)
Comment (brettlangdon) [reply ddtrace/sampler.py:28]: yes, that could be a good follow-up action item. now that we have the benchmark case, could be easy to validate (at least in the context of SamplingRule)

Comment (brettlangdon) [reply ddtrace/sampler.py:28]: With only doing `cache.py` -> `cache.pyx`

```
+------------------------------+---------------------------------------------------------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------+
| Benchmark                    | /artifacts/b615b5ec-01ed-4d07-a939-e325c65a9373/sampling_rule_matches/0.56.0rc4.dev20+gaac21153//results.json | /artifacts/b615b5ec-01ed-4d07-a939-e325c65a9373/sampling_rule_matches/0.56.0rc4.dev21+g2a6f8493//results.json |
+==============================+===============================================================================================================+===============================================================================================================+
| samplingrules-high_match     | 58.8 us                                                                                                       | 43.4 us: 1.36x faster                                                                                         |
+------------------------------+---------------------------------------------------------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------+
| samplingrules-average_match  | 59.7 us                                                                                                       | 44.4 us: 1.35x faster                                                                                         |
+------------------------------+---------------------------------------------------------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------+
| samplingrules-low_match      | 65.9 us                                                                                                       | 44.6 us: 1.48x faster                                                                                         |
+------------------------------+---------------------------------------------------------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------+
| samplingrules-very_low_match | 2.38 ms                                                                                                       | 2.03 ms: 1.17x faster                                                                                         |
+------------------------------+---------------------------------------------------------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------+
| Geometric mean               | (ref)                                                                                                         | 1.33x faster                                                                                                  |
+------------------------------+---------------------------------------------------------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------+
```

But obviously we can improve the performance a ton by getting rid of Python data structures.

Probably not worth adding the overhead of managing something else in Cython right now, unless we get more definitive information that this is a bottleneck in creating traces.

File: ddtrace/sampler.py [hunk 4]
```diff
@@ -396,6 +398,20 @@ def _pattern_matches(self, prop, pattern):
         # Exact match on the values
         return prop == pattern
 
+    @cachedmethod()
+    def _matches(self, key):
+        # type: (Tuple[Optional[str], str]) -> bool
+        service, name = key
+        return all(
```
Comment (Kyle-Verhoog) [inline ddtrace/sampler.py:405]: Could we optimize here by short circuiting? AFAICT this will always loop over every pattern

**Review by Kyle-Verhoog — COMMENTED** [review]
Just one thought, otherwise lgtm

File: ddtrace/sampler.py (hunk 4 above)
Comment (brettlangdon) [reply ddtrace/sampler.py:405]: well, we sort of do in the match pattern function, if no rule is set we return immediately `True`.

we could do that here instead, only call if the pattern is not `NO_RULE`, but /shrug not sure there is much of a difference there.

what do you think? how would you think to optimize further?

Comment (Kyle-Verhoog) [reply ddtrace/sampler.py:405]: I'm thinking to rewrite this as

```
for prop, pattern in [(service, self.service), (name, self.name)]:
    if not self._pattern_matches(prop, pattern):
        return False
else:
    return True
```

File: ddtrace/sampler.py [hunk 5]
```diff
@@ -396,6 +398,20 @@ def _pattern_matches(self, prop, pattern):
...
+            [
+                self._pattern_matches(prop, pattern)
+                for prop, pattern in [
+                    (service, self.service),
+                    (name, self.name),
+                ]
+            ]
+        )
```
Comment (brettlangdon) [inline ddtrace/sampler.py:413]:
```suggestion
    for prop, pattern in [(service, self.service), (name, self.name)]:
        if not self._pattern_matches(prop, pattern):
            return False
    else:
        return True
```

File: ddtrace/sampler.py [hunk 6]
```diff
@@ -396,6 +398,16 @@ def _pattern_matches(self, prop, pattern):
...
 
+    @cachedmethod()
+    def _matches(self, key):
+        # type: (Tuple[Optional[str], str]) -> bool
+        service, name = key
+    for prop, pattern in [(service, self.service), (name, self.name)]:
+        if not self._pattern_matches(prop, pattern):
+            return False
+    else:
+        return True
```
Comment (brettlangdon) [inline ddtrace/sampler.py:409]:
```suggestion
        for prop, pattern in [(service, self.service), (name, self.name)]:
            if not self._pattern_matches(prop, pattern):
                return False
        else:
            return True
```

**Review by Kyle-Verhoog — APPROVED** [review]

**Review by majorgreys — APPROVED** [review]

**Pull request merged**